kvstoremesh: lock destination etcd

balous opened the pull request:
Create per-cluster lock in destination etcd so that multiple instances of kvstoremesh can be used to sync data to a single etcd cluster. Every remote cluster can be synced by a single instance but each instance can sync different remote clusters.

Please ensure your pull request adheres to the following guidelines:

(https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [ ] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [ ] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [ ] Provide a title or release-note blurb suitable for the release notes.
- [x] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)
- [x] Thanks for contributing!

This PR implements local backend locking in kvstoremesh so that multiple instances can be run over a single etcd cluster. Local backend is locked on per-remote-cluster bases as kvstoremesh opens a new etcd connection for each remote cluster. This, in theory, opens ways to more creative approaches like having multiple instances of kvstoremesh syncing different set of clusters (aka sharding) but also opens questions with difficult answers - whether to drain removed cluster if there could possibly be another instance still syncing it.

This PR implements locking within remote cluster's reconnections logic just before retrieving remote configuration. The lock is locked with timeout, which, upon expiration, makes reconnection to fail and go around. It is assumed that the lock can't be lost without losing etcd connection and thus having to reconnect. This way, there is no need to implement shutdown of sync watchers.

There are also unsolved problems:
- status - status can't report that some particular cluster is waiting for lock. It just states that the cluster is ready with outdated numbers of synced resources. A new field could be added to the status indicating locking status, but it would require to change status data structures and I would rather get some commentery for the basic idea before going into this.
- metrics - metrics can indicate two states - cluster is ready or not and I think neither of them is correct for the situation when the cluster is waiting for lock. The `cilium_kvstoremesh_remote_cluster_readiness_status` metrics could be made more general by adding a new label `state` whose value could contain one of [`ready`, `disconnected`, `locked`] but this must be carefuly thought of so that potential upgrade won't trigger alerts for rules like `cilium_kvstoremesh_remote_cluster_readiness_status < 1`.
- removed cluster drainage - kvstoremesh implements cluster draining upon config file removal. The drainage must be done only when the config file was removed from the last instance of kvstoremesh HA setup but the instances are not aware of each other.

Please @giorio94, check my chages and comment. I feel I can't continue with the PR without consultation with kvstoremesh maintainers.

Fixes: #37127

```release-note
KVStoreMesh: add support for leader election, to allow running multiple replicas when Cilium operates in kvstore identity allocation mode.
```

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -134,17 +143,56 @@ func RegisterSyncWaiter(p SyncWaiterParams) {
+			km.common.Stop(km.ctx)
+			lock.Unlock(km.ctx)
+			return
+		// TBD: are we the only reader of the channel?
```
Yes, you are for the local client (at least currently), but this channel does not strictly give you the signal that the lock got lost, as based on other status checks. In turn, this risks causing two instances of kvstoremesh to behave as leaders at the same time, with problematic consequences. 

The correct signal to check for expiration is whether the lease associated with that lock fails to be renewed, which is given here \[1]. This logic already allows to register a callback executed for all keys associated with that lease, but that's currently not configured for the lock leases \[2]. I would suggest looking into how to enable that for locks as well (the simplest approach is probably just mimicking what's already done for the other leases.

\[1]:  https://github.com/cilium/cilium/blob/978486df7460c5e6ab58d83f75f68baee47190cd/pkg/kvstore/etcd_lease.go#L280-L301
\[2]: https://github.com/cilium/cilium/blob/978486df7460c5e6ab58d83f75f68baee47190cd/pkg/kvstore/etcd.go#L504-L505

balous replied:
I added a new method `RegisterLockLeaseExpiredObserver` to `BackendOperations` interface. I've copy-pasted `RegisterLeaseExpiredObserver`, may be I could create a reusable implementation. Is woth it?

giorio94 replied:
As you prefer honestly, both alternatives work for me considering how trivial the implementation is.

balous replied:
OK, I'll leave as it is now.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -134,17 +143,56 @@ func RegisterSyncWaiter(p SyncWaiterParams) {
+			select {
+			case <-km.ctx.Done():
+				return
+			case <-errTimer.C:
```
Nit. This could simply be `<-time.After()`, given that the timer is only used once.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -134,17 +143,56 @@ func RegisterSyncWaiter(p SyncWaiterParams) {
 	)
 }
 
+func (km *KVStoreMesh) run() {
```
I would suggest moving this logic outside of the kvstoremesh cell, following a (simplified version of the) pattern similar to that of the Cilium operator if possible [1,2]. The two main reasons being easier extensibility, to possibly allow other cells to be guarded by the leader election as well (e.g., the heartbeat one), as well as keeping this cell independent of the leader election logic, for easier integration testing.

\[1]: https://github.com/cilium/cilium/blob/978486df7460c5e6ab58d83f75f68baee47190cd/operator/cmd/root.go#L217-L218
\[2]: https://github.com/cilium/cilium/blob/978486df7460c5e6ab58d83f75f68baee47190cd/operator/cmd/root.go#L89-L97

balous replied:
I made similar implementation to the opeator one and it seems to work fine.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -134,17 +143,56 @@ func RegisterSyncWaiter(p SyncWaiterParams) {
+		// TBD: are we the only reader of the channel?
+		case e := <-km.backend.StatusCheckErrors():
+			km.logger.With(e).Info("Synchronization stopped")
+			km.common.Stop(km.ctx)
```
A common approach is to trigger a full restart of the container when leader election is lost. I'd probably do the same here, just to make sure to restart from a clean state (you can use `hive.Shutdowner` for that).

balous replied:
New implementation stops whole process.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -134,17 +143,56 @@ func RegisterSyncWaiter(p SyncWaiterParams) {
+				continue
+			}
+		}
+		km.logger.Info("Locked backend lock")
```
I would suggest adding an `is_leader` metric, to make it easy to monitor that exactly one leader is active at any time.

balous replied:
I am wondering how to do this. Should I mimic the same metric that is produced by kubernetes leader election with different labels? E.g:

```
leader_election_master_status{name="/kvstoremesh-lock"} 1
```

giorio94 replied:
Yeah, mimicking the Kubernetes leader election one makes sense to me. I'd probably make the label value a bit more expressive, but I'm not sure about what is used by K8s normally.

balous replied:
`controller-manager` has this:

```
leader_election_master_status{name="kube-controller-manager"}
```

`name` label contains name of `leases.coordination.k8s.io` k8s object so I guess putting the lock path within the etcd database is appropriate.

giorio94 replied:
I'd suggest simply using `kvstoremesh` as name at that point, given that the lock path is more an implementation detail.

balous replied:
So I implemented it:

```
$ curl -s localhost:9964/metrics | grep leader
# HELP cilium_kvstoremesh_leader_election_master_status leader election status
# TYPE cilium_kvstoremesh_leader_election_master_status gauge
cilium_kvstoremesh_leader_election_master_status{name="kvstoremesh"} 1
```

 There is one imperfection - hive prefixes all metrics with `cilium_kvstoremesh_` thus making it incompatible with similar metric created by leader election k8s library. Are there some posibilities to overcome this?

giorio94 replied:
That is because the metric is configured with the `cilium_kvstoremesh` namespace, otherwise it would not. But it looks reasonable to me, for consistency with all the other kvstoremesh metrics.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -108,7 +117,7 @@ func newKVStoreMesh(lc cell.Lifecycle, params params) *KVStoreMesh {
+	//	lc.Append(km.common)
 
 	return &km
 }
```
Commenting here, as the UI doesn't allow be to select the correct line, but this refers to the `RegisterSyncWaiter` function. Basically, the goal there is have a readiness signal (i.e., kvstoremesh connected to all the remote clusters, and completed the initial synchronization), which is exposed both via a metric and as part of the `/readyz` endpoint \[1], used for the readiness probe.

The goal of the readiness probe in the normal case is to prevent Cilium agents from connecting when synchronization has not completed yet, to be able to use higher bootstrap QPS. However, this does not make sense in this context, and would need to be adapted because otherwise only the leader replica would eventually turn ready, which is problematic. 

Additionally, these changes do break the health check in the normal case, because it is based on the assumption that the initial list of clusters is retrieved synchronously during the execution of the `kc.common` start hook, hence ensuring that `ForEachRemoteCluster` in [2] loops on the populated set of target clusters when it gets invoked. However, that's no longer guaranteed to be the case given that the start hook is executed asynchronously. Migrating to the Cilium operator pattern as suggested above should help here as well.

\[1] https://github.com/cilium/cilium/blob/978486df7460c5e6ab58d83f75f68baee47190cd/clustermesh-apiserver/kvstoremesh/health.go#L32-L51
\[2] https://github.com/cilium/cilium/blob/978486df7460c5e6ab58d83f75f68baee47190cd/pkg/clustermesh/kvstoremesh/kvstoremesh.go#L194-L198

balous replied:
After leader election implementation I found that things didn't improve. `synced` function is executed sooner than the files are loaded. So I implemented `common.WaitLoaded()` function that blocks until all preexisting files are loaded.

balous replied:
I am unsure about this solution. `synced()` function is triggered by a hive's `Invoke()` function which doesn't seem to obey the leader ifecycle I implemented. Is there some way to acheieve this with hive or do we need to implement waiting like this?

giorio94 replied:
Hmm, that seems a quirk associated with the functioning of `WithLeaderLifecycle`. I would try with something like this:

```diff
diff --git a/clustermesh-apiserver/kvstoremesh/lifecycle.go b/clustermesh-apiserver/kvstoremesh/lifecycle.go
index ccd30e2a7a95..ca7f4831c6bd 100644
--- a/clustermesh-apiserver/kvstoremesh/lifecycle.go
+++ b/clustermesh-apiserver/kvstoremesh/lifecycle.go
@@ -4,7 +4,11 @@
 package kvstoremesh
 
 import (
+	"log/slog"
+	"runtime/pprof"
+
 	"github.com/cilium/hive/cell"
+	"github.com/cilium/hive/job"
 )
 
 // LeaderLifecycle is the inner lifecycle of the kvstoremesh that is started when this
@@ -22,8 +26,10 @@ func WithLeaderLifecycle(cells ...cell.Cell) cell.Cell {
 			func() *LeaderLifecycle { return &LeaderLifecycle{} },
 		),
 		cell.Decorate(
-			func(lc *LeaderLifecycle) cell.Lifecycle {
-				return lc
+			func(reg job.Registry, h cell.Health, l *slog.Logger, llc *LeaderLifecycle, mid cell.ModuleID) (cell.Lifecycle, job.Group) {
+				return llc, reg.NewGroup(h, llc,
+					job.WithLogger(l),
+					job.WithPprofLabels(pprof.Labels("cell", string(mid))))
 			},
 			cells...,
 		),
```

This bit would still need to be taken care of though:

> would need to be adapted because otherwise only the leader replica would eventually turn ready, which is problematic.

One possibility could be marking the replica as ready once the leader election process if attempted once and it returned an error due to another one already holding the lock. That will require a couple of adaptations to `RegisterSyncWaiter` though.

balous replied:
In [another thread](https://github.com/cilium/cilium/pull/39848#discussion_r2190920559), we've aggreed to make the timeout for `lockPath()` configurable and infinite for leader election, so making the service ready after first timeout won't work. In addition, it would be easy to errorneously configure the probe with too short timeout in the chart.

So what about making it ready after successfuly connecting to backend?

giorio94 replied:
> In https://github.com/cilium/cilium/pull/39848#discussion_r2190920559, we've agreed to make the timeout for lockPath() configurable and infinite for leader election, so making the service ready after first timeout won't work.

That would be fairly easy to solve by performing a first attempt with a relatively short timeout (i.e., a few seconds) to verify if we can acquire the lock immediately, falling back to an infinite wait otherwise.

> In addition, it would be easy to errorneously configure the probe with too short timeout in the chart.

Not sure I'm understanding what you mean here.

> So what about making it ready after successfuly connecting to backend?

Nope, that would not work in the general case (etcd as sidecar), because it would break the assumptions that I mentioned at the beginning of this thread. But one possibility could be making the probe behavior configurable via a parameter, so that it is possible to use the correct one depending on whether kvstoremesh is run with a sidecar etcd or targeting an external etcd cluster.

balous replied:
So I implemented double locking with 10s and infinite timeouts. There was one problem - `SyncState` module schedules its `Stop` function via hive. As leader election runs on background, I had to [reimplement](https://github.com/cilium/cilium/pull/39848/commits/f7600bf4e05a71c1d681d693740fdfd6e01db523) it to keep track of running routines via a counter. Stop function is called when this counter first reaches 0.

giorio94 replied:
The new `SyncState` approach does not look correct to me, but I've commented directly there, so marking this thread as resolved.

giorio94 reviewed (commented):
Thanks, and sorry for the delayed review.

Yeah, I agree that this looks way more straightforward. I've left a few comments inline about high-level aspects that IMO should be improved. (A rebase is also necessary, as the etcd bootstrap logic changed a bit recently --- most notably, the client promise has been dropped in favor of a plain client object, adapting kvstoremesh as well).

balous committed:
kvstore: implemented lock lease expiration observer

This commit implements possibility to register lock expiration observer in kvstore backend.

Signed-off-by: Petr Baloun <petr.baloun@firma.seznam.cz>

giorio94 commented on pkg/kvstore/memory.go:
```diff
@@ -205,6 +205,11 @@ func (c *inMemoryClient) LockPath(ctx context.Context, path string) (KVLocker, e
 
+// RegisterLockLeaseExpiredObserver implements BackendOperations.
+func (c *inMemoryClient) RegisterLockLeaseExpiredObserver(prefix string, fn func(key string)) {
+	panic("unimplemented")
```
Nit. This method should perform nothing (as the one above), rather than panic'ing.

balous replied:
`panic` has been removed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +54,60 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	})
+
+	for {
+		log.Info("Locking backend")
```
Nit. `log.Info("Acquiring leader election lock")`. And similarly for the other log lines, so that it is clearer to users what is happening.

balous replied:
Reworded.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +54,60 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+		log.Info("Locking backend")
+		lock, err := client.LockPath(ctx, backendLockName)
+		if err != nil {
+			log.With("error", err).Warn("Failed to acquire backend lock")
```
It is expected to see an error here if another replica is already the leader, isn't it? In that case, the error should not be emitted at warning level.

balous replied:
There is a problem with this. `LockPath()` has an [hardcoded timeout](https://github.com/cilium/cilium/blob/main/pkg/kvstore/etcd.go#L618) which makes it to exit after waiting a minute. Making this error would cause errors to appear in the log during normal operation.

What to do with this - distinguish between returned errors or make the timeout of the function configurable (and set it to infinite for this call)?

giorio94 replied:
Moving the timeout to the callers makes sense to me, given that this function already takes a context as parameter. It should be also a fairly small change, considering that it is only called in a couple of places excluding tests.

balous replied:
Timeout on the caller side: https://github.com/cilium/cilium/pull/39848/commits/93f46fda76743c2b271486a742132b1ac05f2628

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +54,60 @@ func NewCmd(h *hive.Hive) *cobra.Command {
 	return rootCmd
 }
+
+func registerKVStoreMeshHooks(log *slog.Logger, lc cell.Lifecycle, llc *LeaderLifecycle, client kvstore.Client, shutdowner hive.Shutdowner) {
```
Nit. How about `registerLeaderElectionHooks`? That makes the purpose of this function a bit clearer IMO (and similarly for the function below).

balous replied:
Renamed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +54,60 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+func runKVStoreMesh(ctx context.Context, log *slog.Logger, lc *LeaderLifecycle, client kvstore.Client, shutdowner hive.Shutdowner) {
+	client.RegisterLockLeaseExpiredObserver(backendLockName, func(string) {
+		log.Error("Lost backend lock.")
+		shutdowner.Shutdown()
```
Nit. I'd suggest passing an error as well, to make the shutdown reason clear: `hive.ShutdownWithError(errors.New("Leader election lost"))`

balous replied:
Error added.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +54,60 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+			}()
+			return nil
+		},
+		OnStop: func(ctx cell.HookContext) error {
```
Nit. `s/ctx/hctx`, just for extra clarity.

balous replied:
Renamed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +54,60 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+		if err != nil {
+			log.With("error", err).Error("Failed to unlock backend lock")
+		}
+		return
```
Nit. This return is redundant.

balous replied:
Removed.

giorio94 reviewed (commented):
Thanks! A few comments inline, but overall the approach makes sense to me.

giorio94 commented on pkg/kvstore/memory.go:
```diff
@@ -206,9 +206,7 @@ func (c *inMemoryClient) RegisterLeaseExpiredObserver(prefix string, fn func(key
-func (c *inMemoryClient) RegisterLockLeaseExpiredObserver(prefix string, fn func(key string)) {
-	panic("unimplemented")
-}
+func (c *inMemoryClient) RegisterLockLeaseExpiredObserver(prefix string, fn func(key string)) {}
```
Nit. Should be moved to the first commit.

balous replied:
Moved.

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -614,9 +614,7 @@ func (e *etcdClient) newExpBackoffRateLimiter(name string) backoff.Exponential {
-	// Create the context first, so that the timeout also accounts for the time
-	// possibly required to acquire a new session (if not already established).
-	ctx, cancel := context.WithTimeout(ctx, time.Minute)
+	ctx, cancel := context.WithCancel(ctx)
```
This line (and the one below) can now be simply deleted.

balous replied:
Removed.

giorio94 commented on pkg/kvstore/allocator/allocator_test.go:
```diff
@@ -120,7 +120,9 @@ func BenchmarkRunLocksGC(b *testing.B) {
 			kvstore.EtcdAddrOption: kvstore.EtcdDummyAddress(),
 		}, kvstore.ExtraOptions{})
-		lock2, err = client.LockPath(context.Background(), allocatorName+"/locks/"+shortKey.GetKey())
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
I guess this should be `time.Minute` to preserve the previous behavior (or otherwise the timeout could be just dropped).

balous replied:
Removed with the rest of test updates.

giorio94 commented on pkg/kvstore/etcd_test.go:
```diff
@@ -84,7 +84,9 @@ func TestGetIfLocked(t *testing.T) {
 				key := randomPath + "foo"
-				kvlocker, err := client.LockPath(context.Background(), "locks/"+key+"/.lock")
+				ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+				defer cancel()
```
I'd be personally fine to revert these changes (same for all the other occurrences in this file) for simplicity, and simply let the test timeout kick in in case there's something wrong and the lock is not acquired in time.

balous replied:
OK, removed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/cells.go:
```diff
@@ -27,18 +27,28 @@ var Cell = cell.Module(
+	// for leader election is guaranteed to be executed last, when
+	// all the previous ones have already completed. Otherwise, cells within
+	// the "WithLeaderLifecycle" scope may be incorrectly started too early,
+	// given that "registerKVStoreMeshHooks" does not depend on all of their
```
Nit. `s/registerKVStoreMeshHooks/registerLeaderElectionHooks`

balous replied:
Fixed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/lifecycle.go:
```diff
@@ -0,0 +1,37 @@
+func WithLeaderLifecycle(cells ...cell.Cell) cell.Cell {
+	return cell.Module(
+		"leader-lifecycle",
+		"Kvstoremesh Leader Lifecycle",
```
Nit. `s/Kvstoremesh/KVStoreMesh/`

balous replied:
Fixed.

giorio94 commented on pkg/clustermesh/kvstoremesh/cell.go:
```diff
@@ -23,4 +23,5 @@ var Cell = cell.Module(
 	cell.Config(common.DefaultConfig),
 
 	metrics.Metric(common.MetricsProvider("")),
+	metrics.Metric(MetricsProvider("")),
```
Nit. The subsystem parameter is always empty, so it can be simply dropped.

balous replied:
Removed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +59,85 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	kvMetrics.LeaderElectionStatus.With(prometheus.Labels{metrics.LabelLeaderElectionName: "kvstoremesh"}).Set(float64(0))
+
+	log.Info("Acquiring leader election lock")
+	toCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
```
Nit. It would be valuable to add a comment explaining the reason for this two-stage approach.

balous replied:
Added.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +59,85 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	if err == nil {
+		syncedCallback(ctx)
+	} else {
+		if err.Error() == "etcd client timeout exceeded" {
```
I'd suggest defining this as an error variable, to avoid having to match on strings, which is always fragile.

balous replied:
The problem here is that etcd client library doesn't define the error itself. Declaring the string as constant in cilium code won't improve the situation.

giorio94 replied:
That error is defined in the Cilium codebase actually: https://github.com/cilium/cilium/blob/daa9fa24b9d97e9992b7676af597b07e255751d7/pkg/kvstore/etcd.go#L296-L302

balous replied:
Yes, it is, but as a mere string not as type like `context.DeadlineExceeded` is. It is not possible to use `errors.Is()` to compare the error, all I can do is to compare the message. Or is there some better possiblity I am unaware of?

giorio94 replied:
Yes, that's why I was suggesting to extract it into a variable.

balous replied:
OK, added a string constant for the error message.

giorio94 replied:
Why not directly defining the error as a variable, so that you can directly match on it with `errors.Is`?

balous replied:
I tried `errors.Is(err, fmt.Errorf(etcdTimeoutErr))` but it doesn't work. The strings are the same but they are not the same variable so the `errors.Is()` function won't regard the two errors as the same error.

giorio94 replied:
Yes, the error should be extracted into a variable in `etcd.go`, rather than generating a new one every time. And then just compare with it using `errors.Is()`

balous replied:
I added `EtcdTimeoutError` which is returned by `Hint()` function instead of `fmt.Errorf('etcd timeout exceeded')`. Sorry for not figuring this quicky, I mistakenly though that the error is generated in etcd client itself, not in Cilium's `kvstore`.

giorio94 replied:
No worries, that part of the codebase is definitely not easy to navigate at the beginning. One more simplification comment inline.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +59,85 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	})
+}
+
+func runLeaderElection(ctx context.Context, log *slog.Logger, lc *LeaderLifecycle, client kvstore.Client, shutdowner hive.Shutdowner, kvMetrics kvstoremesh.Metrics, SyncState syncstate.SyncState) {
```
Nit. I'd suggest simply passing the params struct, to reduce the number of parameters.

balous replied:
Reworked.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +59,85 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	log.Info("Leader election lock acquired")
+	kvMetrics.LeaderElectionStatus.With(prometheus.Labels{metrics.LabelLeaderElectionName: "kvstoremesh"}).Set(float64(1))
+
+	lc.Start(log, ctx)
```
The error should not be ignored.

balous replied:
Added handling.

giorio94 replied:
Should also call `params.Shutdowner.Shutdown` to let the process terminate.

balous replied:
Added.

giorio94 commented on pkg/clustermesh/kvstoremesh/kvstoremesh.go:
```diff
@@ -122,16 +122,6 @@ type SyncWaiterParams struct {
-		}),
-	)
-}
-
```
This should not be removed.

balous replied:
This was replaced by the call from `runLeaderElection()`.

giorio94 replied:
That's not correct though, as it modifies the behavior. We still want to only mark the replica that can acquire the lock immediately as ready only after that synchronization completed, as we did previously.

balous replied:
Now I see. So I reworked the leader election to call `RegisterSyncWaiter()` directly outside of hive. It also calls `SyncState.Stop()` directly.

giorio94 commented on clustermesh-apiserver/syncstate/syncstate.go:
The new approach is not correct, because SyncState may be marked ready too early if not all resources have registered before all the others are marked as complete. I would suggest dropping the start hook as you did, and then calling `ss.Stop()` where appropriate to preserve the existing behavior.

balous replied:
Basically, it works in kvstoremesh but I have to figure out how to call it in clustermesh-apiserver.

balous replied:
So I added a start hook to `clustermesh-synchronization` cell that calls it.

giorio94 reviewed (commented):
Thanks for the updates. A few more comments inline.

balous committed:
kvstoremesh: leader election with etcd lock
Fixes: #37127

Create lock in destination etcd so that multiple instances
of kvstoremesh can be used to sync data.

Signed-off-by: Petr Baloun <petr.baloun@firma.seznam.cz>

balous committed:
kvstore: moved LockPath timeout to the caller

Caller is now responsible to set the context timeout so that the timeout length is not hardcoded inside the function.

Signed-off-by: Petr Baloun <petr.baloun@firma.seznam.cz>

balous committed:
clustermesh-apiserver: moved syncstate OnStart hook outside of syncstate

kvstoremesh leader election runs outside of hive and uses syncstate
to indicate its status to the rest of the service. So syncstate
must not to rely on hive.

Signed-off-by: Petr Baloun <petr.baloun@firma.seznam.cz>

giorio94 commented on clustermesh-apiserver/clustermesh/cells.go:
```diff
@@ -120,8 +120,20 @@ var Synchronization = cell.Module(
 		),
 		cell.Invoke(RegisterSynchronizer[*cilium_api_v2a1.CiliumEndpointSlice]),
 	),
+	cell.Invoke(registerSyncStateStop),
```
Nit. The commit message should be updated with this new approach.

balous replied:
Reworded.

giorio94 commented on pkg/clustermesh/kvstoremesh/metrics.go:
```diff
@@ -0,0 +1,27 @@
+	LeaderElectionStatus metric.Vec[metric.Gauge]
+}
+
+func MetricsProvider() func() Metrics {
```
Nit. No need for returning a function, you can directly return the `Metrics` object.

balous replied:
Reworked.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +63,93 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+
+	params.Metrics.LeaderElectionStatus.With(prometheus.Labels{metrics.LabelLeaderElectionName: "kvstoremesh"}).Set(float64(0))
+
+	// try to win leader election with short timeout to test backend function
```
I'd replace `test backend function` with something like `to verify if we can be immediately promoted as leader (e.g., we are the only replica)` for increased clarity.

balous replied:
Reworded.

balous committed:
kvstore: return specific error instance in case of etcd client timeout

kvstore functions return EtcdTimeoutError instead of fmt.Errorf()
so that it is possible to test for the error with errors.Is().

Signed-off-by: Petr Baloun <petr.baloun@firma.seznam.cz>

giorio94 commented on pkg/kvstore/etcd.go:
```diff
@@ -94,6 +94,13 @@ var (
+
+var EtcdTimeoutError error = timeoutError{}
+
+func (e timeoutError) Error() string { return "etcd client timeout exceeded" }
```
Can be simplified with (IIRC the linter expects the name to start with `Err`):

```
var ErrEtcdClientTimeout = errors.New("etcd client timeout exceeded")
```

balous replied:
Renamed and reimplemented.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,106 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	// with infinite timeout.
+	params.Log.Info("Acquiring leader election lock")
+	toCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+	defer cancel()
```
Nit. Cancel should be called after `params.Client.LockPath` terminates, as deferring it here means it would never be executed normally.

balous replied:
Fixed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,106 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+		// signal readiness
+		syncedCallback := params.SyncState.WaitForResource()
+		params.SyncState.Stop()
+		syncedCallback(ctx)
```
Nit. Calling `Stop` alone should be enough, without the need for Wait/callback.

balous replied:
When I remove `syncedCallback(ctx)` readiness probe won't never succeed.

If I understand [the code](https://github.com/cilium/cilium/blob/main/pkg/lock/stoppable_waitgroup.go#L32) correctly, calling `Stop()` causes `Add()` to do nothing, but `Wait()` still waits for all the routines that have registered.

giorio94 replied:
Sure, I meant dropping the `WaitForResource` call as well (as it is currently registered and then immediately marked as done).

balous replied:
Yes, it did the trick.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,106 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	<-ctx.Done()
+
+	params.Log.Info("Releasing leader election lock")
+	err = lock.Unlock(context.Background())
```
One tricky aspect is that when running with etcd as a sidecar, it is possible that, when the pod is restarted, etcd stops before kvstoremesh, which leads to this call to Unlock being blocking. It'd suggest setting a tight timeout (e.g., 1 or 2 seconds) to make sure we don't unnecessarily block for too long. I'd probably also convert the Error log into an info one, given that it is actually expected to happen in that case.

balous replied:
Added 2 seconds timeout.

giorio94 replied:
Thanks. It would be nice to add a comment for future reference as well.

balous replied:
Added.

giorio94 reviewed (commented):
Thanks, overall makes sense to me. Just a couple of nits inline, I'll take a final look once marked ready for review.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,109 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	params.Log.Info("Acquiring leader election lock")
+	toCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+
+	var lock kvstore.KVLocker
```
Nit. Redundant, as you are using `:=` in the assignment below.

balous replied:
Removed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,109 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+
+	err = lc.Start(params.Log, ctx)
+	if err != nil {
+		params.Log.Error("Failed to run KvStoremesh", logfields.Error, err)
```
Nit. `s/KvStoremesh/KVStoreMesh`

balous replied:
Fixed.

giorio94 commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,109 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	params.Log.Info("Leader election lock acquired")
+	params.Metrics.LeaderElectionStatus.With(prometheus.Labels{metrics.LabelLeaderElectionName: "kvstoremesh"}).Set(float64(1))
+
+	kvstoremesh.RegisterSyncWaiter(kvstoremesh.SyncWaiterParams{
+		KVStoreMesh: params.KVStoreMesh,
+		SyncState:   params.SyncState,
+		Lifecycle:   lc,
+		JobGroup:    params.JobGroup,
+		Health:      params.Health,
+	})
```
The reason why this works is a bit :magic_wand:, and heavily relies on Hive internals. We could make '[KVStoreMesh.synced] explicitly block until its start hook has terminated as an extra guarantee, but I wouldn't block this PR on that, given that is has already undergone multiple iterations, and as to be fair that was already the case before.

giorio94 reviewed (approved):
Thanks, looks good to me on my side.

@MrFreezeex, @marseel Please also take an in-depth look, given the potential for regressions to standard kvstoremesh deployments as well (I haven't really convinced myself about the opportunity of hiding this behind a feature flag though, as it would introduce different bootstrap behaviors, and make the logic even more convoluted). On the bright side, we are just at the beginning of the 1.19 cycle, so we have plenty of time to discover and fix possible problems.

marseel commented on pkg/clustermesh/kvstoremesh/metrics.go:
```diff
@@ -0,0 +1,25 @@
+		LeaderElectionStatus: metric.NewGaugeVec(metric.GaugeOpts{
+			Namespace: metrics.Namespace,
+			Subsystem: "",
+			Name:      "leader_election_master_status",
```
nit: no need for "master" in the metric name, it's already about leader :)

balous replied:
Well, I copied the name from k8s leader election. I think it's practical to keep it the same.

derailed reviewed (approved):
@balous Thank you for this update!

MrFreezeex commented on clustermesh-apiserver/kvstoremesh/root.go:
```diff
@@ -44,3 +60,107 @@ func NewCmd(h *hive.Hive) *cobra.Command {
+	err = lock.Unlock(toCtx)
+	if err != nil {
+		params.Log.Error("Failed to release leader election lock", logfields.Error, err)
+	}
```
Shouldn't this be in in a whole defer function right after we acquire the lock? For instance the return in the current line 152 would not execute this AFAICT is that right?

balous replied:
You're right, fixed.

MrFreezeex reviewed (commented):
Thanks! I left a question inline that may require a tiny fix, otherwise lgtm
